language/move-vm/integration-tests/src/tests/bad_storage_tests.rs

```diff
@@ -0,0 +1,551 @@
+
+        let mut blob = vec![];
+        m.serialize(&mut blob).unwrap();
+        storage.publish_or_overwrite_module(m.self_id(), blob);
+
+        let vm = MoveVM::new();
+        let mut sess = vm.new_session(&storage);
+
+        sess.execute_function(
+            &module_id,
+            &fun_name,
+            vec![],
+            vec![],
+            TEST_ADDR,
+            &mut cost_strategy,
+            |x| x,
+        )
+        .unwrap();
+    }
+
+    // Erase the body of M::foo to make it fail verification.
+    // Publish this modified version of M and the VM should fail to load it.
+    {
+        let mut storage = InMemoryStorage::new();
+
+        let mut m = m.into_inner();
+        m.function_defs[0].code.as_mut().unwrap().code = vec![];
+        let m = m.freeze().unwrap();
+        let mut blob = vec![];
+        m.serialize(&mut blob).unwrap();
+        storage.publish_or_overwrite_module(m.self_id(), blob);
+
+        let vm = MoveVM::new();
+        let mut sess = vm.new_session(&storage);
+
+        let err = sess
+            .execute_function(
+                &module_id,
+                &fun_name,
+                vec![],
+                vec![],
+                TEST_ADDR,
+                &mut cost_strategy,
+                |x| x,
+            )
+            .unwrap_err();
+
+        assert!(err.status_type() == StatusType::InvariantViolation);
+    }
+}
+
+#[test]
+fn test_missing_module_dependency() {
+    // Compile two modules M, N where N depends on M.
+    let code = r#"
+        module M {
+            public fun foo() {}
+        }
+
+        module N {
+            use {{ADDR}}::M;
+
+            public fun bar() { M::foo(); }
+        }
+    "#;
+    let code = code.replace("{{ADDR}}", &format!("0x{}", TEST_ADDR.to_string()));
+    let mut units = compile_units(TEST_ADDR, &code).unwrap();
+    let n = as_module(units.pop().unwrap());
+    let m = as_module(units.pop().unwrap());
+
+    let mut blob_m = vec![];
+    m.serialize(&mut blob_m).unwrap();
+    let mut blob_n = vec![];
+    n.serialize(&mut blob_n).unwrap();
+
+    let cost_table = zero_cost_schedule();
+    let mut cost_strategy = CostStrategy::system(&cost_table, GasUnits::new(0));
+
+    let module_id = ModuleId::new(TEST_ADDR, Identifier::new("N").unwrap());
+    let fun_name = Identifier::new("bar").unwrap();
+
+    // Publish M and N and call N::bar. Everything should work.
+    {
+        let mut storage = InMemoryStorage::new();
+
+        storage.publish_or_overwrite_module(m.self_id(), blob_m);
+        storage.publish_or_overwrite_module(n.self_id(), blob_n.clone());
+
+        let vm = MoveVM::new();
+        let mut sess = vm.new_session(&storage);
+
+        sess.execute_function(
+            &module_id,
+            &fun_name,
+            vec![],
+            vec![],
+            TEST_ADDR,
+            &mut cost_strategy,
+            |x| x,
+        )
+        .unwrap();
+    }
```

Comment:
There's quite a bit of copy-n-pasting going on here and I'm aware of it... For now I'd prefer not trying to make them look pretty/optimized as they'll likely be rewritten in a much more concise way once the new YAML based functional tests come.

Reply:
sounds good and no problem with it